cvcdesc.1
.\" Copyright (c) 2004 Specifix, Inc.
.TH CVCDESC 1 "2 November 2004" "Specifix, Inc."
.SH NAME
cvcdesc \- Conary trove metadata XML generator
.SH SYNOPSIS
.B cvcdesc \fI<options> \fR[\fIoutput-file-name\fR]
.SH DESCRIPTION
\fBcvcdesc\fR generates an XML file containing metadata information
suitable for \fBcvc describe\fR. It can retrieve metadata from several
sources using \fI--source\fP or manually from options specified on the
command line.
.SH OPTIONS 
.TP
.B \-\-source \fIfreshmeat|...\fP
Specify an external source. 'freshmeat' is the only currently supported source.
.TP
.B \-\-pkg-name \fIpackage-name\fP
Query the external source for \fIpackage-name\fP.
.TP
.B \-\-short-desc \fI<string>\fP and \-\-long-desc \fI<string>\fP
Specify the package's short description and long description, respectively.
Short description is required, long description, and all other tags, are
optional.
.TP
.B \-\-url <string>
Specify a package URL. More that one URL may be specified on the command line.
.TP
.B \-\-category <string>
Specify a package category. More than one category may be specified.
.TP
.B \-\-license <string>
Specify a package license. More than one license may be specified.
.\"
.SH EXAMPLES
cvcdesc --source=freshmeat --pkg-name a2ps a2ps.xml

cvcdesc --short-desc "A short description" simple.xml
.\"
.SH "SEE ALSO"
cvc(1), conary(1)


NEWS
Changes in 2.5.7:
  o New Features:
    * Attempt to restart an interrupted download instead of raising a
      'changeset was truncated in transit' error. The newly added
      downloadAttempts, downloadRetryThreshold, and downloadRetryTrim
      configuration options govern this behavior. (CNY-3886)
    * Add config options for selecting default group and factory recipe
      templates (CNY-3887)
    * Improve update speed for troves with a large quantity of unchanged
      files.
    * Mirroring to multiple targets now operates concurrently across all
      targets rather than sequentially.

  o Bug Fixes:
    * Fixed crash when using saved passwords after the session keyring
      was revoked. (CNY-3858)
    * Cached passwords are saved in the per-user keyring instead of the
      per-session keyring. (CNY-3858)
    * Introduce versioned RPM dependencies for any require or provide
      that has an exact version. This behavior can be disabled by
      setting the 'enableRPMVerisonDeps' config option to False.
      (CNY-3882)
    * Fixed capsule postinstall scripts failing if they referenced a
      shared file being preserved during the same job. For example,
      installing a library that calls ldconfig in the same job as
      installing glibc.i686 alongside glibc.x86_64. (CNY-3884)
    * Fixed updating encapsulated and native packages which move a file
      across a symlinked directory, especially in the case of UsrMove.
      (CNY-3885)
    * Fixed error when updating binaries with out-of-date prelinks.

  o Internal Changes:
    * The makefile has been updated to avoid running kidc when running
      'make install' with 'NO_KID=1'. (CNY-3887)

Changes in 2.5.6:
  o Bug Fixes:
    * Interrupting an update with Ctrl-C during "Committing database
      transaction" no longer results in the filesystem journal being
      rolled back after the database has already committed. (CNY-3106)
    * Fixed getChangeSetSize failing when given a job containing
      erasures. (CNY-3876)
    * Fixed a "pathId not in changeset" crash when updating a trove
      across labels after local filesystem file types were changed.
      (CNY-3877)
    * Allow None types when serializing update jobs. (CNY-3879)
    * Fixed a race condition in standalone proxy caches that could cause
      a crash when three or more clients requested the same changeset
      simultaneously. (CNY-3880)
    * Fixed "conary repair" crashing when a config file was replaced by
      a symlink. (CNY-3881)

  o Internal Changes:
    * Committing to repositories running Conary older than 2.1.0 is no
      longer supported.
    * Removed support for mod_python in the repository. Use a WSGI
      server such as gunicorn or uwsgi.

Changes in 2.5.5:
  o New Features:
    * Added a new repository content store type 'flat'. In this
      configuration contents are stored without any intermediate
      directories, which may improve access times and will improve
      sequential access by e.g. backups. To use it, use the syntax
      "contentsDir flat /path/to/store". (CNY-2406)
    * The time at which a trove was committed is now shown in "conary rq
      --debug --info". It is also shown without --debug for source
      troves. (CNY-3487)
    * Added a new 'subPackages' troveinfo that enumerates the names of
      packages created via PackageSpec and groups created via
      createGroup from the same source. Query, repquery, and
      showchangeset commands with the --info and --debug flag set will
      display the list of subpackages.

  o Bug Fixes:
    * The online help for GroupSetRecipe has been fixed to refer to the
      CML method that exists, rather than the SystemModel method that
      does not. (CNY-3868)
    * Fixed crash in GroupSetRecipe when using an empty find()
      (CNY-3872)

  o Internal Changes:
    * When committing changesets to a repository, regular contents are
      written before the database operations begin. This avoids holding
      the commit lock any longer than necessary.
    * Upgraded the embedded SQLite version to 3.8.7.1

Changes in 2.5.4:
  o New Features:
    * Progress information is no longer emitted when downloading
      ephemeral sources during a cook. This keeps logs from being
      bloated when using very large sources. (CNY-3865)
    * Add createAppOnlyArchive script for creating archives that could
      be used in an App Only deployment.
    * r.MakeInstall now takes a makeName argument for overriding the
      name of the make binary. Additionally, a makeName macro may be
      specified to override the default make binary used for both r.Make
      and r.MakeInstall.
    * Added cscache hit/miss counters to the proxy.
    * Minor performance improvements for updates.
    * Add an architecture configuration for i86pc (Solaris x86)
    * Allow tar binary name to be overriden by the tar macro.

  o Bug Fixes:
    * Conary no longer fails to load user/group information on platforms
      where uid 0 or gid 0 are not root. (CNY-3867)
    * Fixed applying updates with a configuration of 'dbPath :memory:'
      always failing with a permission error.
    * Policy no longer fails a build if /etc/ld.so.conf doesn't exist.

  o Internal Changes:
    * When running the test suite, if the python executable being used
      does not have cap_sys_chroot, skip the capsule tests that require
      cap_sys_chroot and print a message documenting how to enable those
      tests.
    * Define a MIN C macro for portabilty to Solaris.
    * Switch install, sed, cp, and make calls to use Makefile macros for
      portabilty to platforms that rename the GNU versions of those
      utilities.
    * Switch to using os.setpgid when configuring the build process
      instead of os.setpgrp because setpgrp creates a new session on
      Solaris.

Changes in 2.5.3:
  o New Features:
    * Key/Value metadata is no longer copied forward during the build

conary/cvcdesc.py
# -*- mode: python -*-
#
# Copyright (c) 2004-2008 rPath, Inc.
#
# This program is distributed under the terms of the Common Public License,
# version 1.0. A copy of this license should have been distributed with this
# source file in a file called LICENSE. If it is not present, the license
# is always available at http://www.rpath.com/permanent/licenses/CPL-1.0.
#
# This program is distributed in the hope that it will be useful, but
# without any warranty; without even the implied warranty of merchantability
# or fitness for a particular purpose. See the Common Public License for
# full details.
#
"""
Generates an XML trove description file.
"""

import sys

from conary import conarycfg
from conary import constants
from conary import metadata
from conary import versions
from conary.lib import log
from conary.lib import options
from conary.lib import util

sys.excepthook = util.genExcepthook()
def usage(rc = 1):
    print "usage: cvcdesc --source <freshmeat|...> --pkg-name <package name> <outputFileName>"
    print "                          or"
    print "               --short-desc <string>"
    print "               [--long-desc <string>]"
    print "               [--url <url>]+"
    print "               [--license <license>]+"
    print "               [--category <category>]+"
    print "               [--language <language code>]+"
    print "               <outputFileName>"
    
    return rc

def realMain(cfg, argv=sys.argv):
    argDef = {}
    cfgMap = {}

    (NO_PARAM,  ONE_PARAM)  = (options.NO_PARAM, options.ONE_PARAM)
    (OPT_PARAM, MULT_PARAM) = (options.OPT_PARAM, options.MULT_PARAM)

    argDef["source"] = ONE_PARAM
    argDef["pkg-name"] = ONE_PARAM
    argDef["long-desc"] = ONE_PARAM
    argDef["short-desc"] = ONE_PARAM
    argDef["url"] = MULT_PARAM
    argDef["license"] = MULT_PARAM
    argDef["category"] = MULT_PARAM

    argDef.update(argDef)

    try:
        argSet, otherArgs = options.processArgs(argDef, cfgMap, cfg, usage,
                                                argv=argv)
    except options.OptionError, e:
        sys.exit(e.val)
    except versions.ParseError, e:
        print >> sys.stderr, e
        sys.exit(1)

    if argSet.has_key('version'):
        print constants.version
        sys.exit(0)

    args = otherArgs[1:]
    log.setVerbosity(1)

    if len(args) != 1:
        return usage()

    longDesc = ""
    urls = []
    categories = []
    licenses = []
    source = "local"
    language = "C"
    
    outFile = args[0]

    if "source" in argSet:
        if argSet["source"] == "freshmeat":
            if "pkg-name" not in argSet:
                log.info("package name required when using external source")
            else:
                fmName = argSet["pkg-name"]

            log.info("fetching metadata from %s", argSet["source"])

            try:
                md = metadata.fetchFreshmeat(fmName)
            except metadata.NoFreshmeatRecord:
                log.error("no freshmeat record found for %s", fmName)
                return 1
            log.info("found record: '%s'", md.getShortDesc())
        else:
            log.error("unsupported metadata source: %s", argSet["source"])
            return 1

        shortDesc = md.getShortDesc()
        longDesc = md.getLongDesc()
        urls = md.getUrls()
        categories = md.getCategories()
        licenses = md.getLicenses()
        source = md.getSource()
        language = "C"
    else:
        if "short-desc" in argSet:
            shortDesc = argSet["short-desc"]
        else:
            log.error("short description must be specified")
            return 0
        
        if "long-desc" in argSet:   longDesc = argSet["long-desc"]
        if "url" in argSet:         urls = argSet["url"]
        if "category" in argSet:    categories = argSet["category"]
        if "license" in argSet:     licenses = argSet["license"]
        if "language" in argSet:    language = argSet["language"]

    xml = """<trove>
    <shortDesc>%s</shortDesc>
    <longDesc>%s</longDesc>
    <source>%s</source>
    <language>%s</language>
""" % (shortDesc, longDesc, source, language)

    for url in urls:
        xml += "    <url>%s</url>\n" % url
    for license_ in licenses:
        xml += "    <license>%s</license>\n" % license_
    for category in categories:
        xml += "    <category>%s</category>\n" % category

    xml += "</trove>\n"

    f = open(outFile, "w")
    f.write(xml)
    f.close()
    log.info("metadata written to %s", outFile)
    return 0

def main(argv=sys.argv):
    try:
        if '--skip-default-config' in argv:
            argv = argv[:]
            argv.remove('--skip-default-config')
            cfg = conarycfg.ConaryConfiguration()
        else:
            cfg = conarycfg.ConaryConfiguration(readConfigFiles=True)
        # reset the excepthook (using cfg values for exception settings)
        sys.excepthook = util.genExcepthook(debug=cfg.debugExceptions)
	return realMain(cfg, argv)
    except conarycfg.CfgError, e:
        log.error(str(e))
        sys.exit(1)

if __name__ == "__main__":
    sys.exit(main())
